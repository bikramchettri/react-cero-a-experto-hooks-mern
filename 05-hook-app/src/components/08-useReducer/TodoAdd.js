import React from 'react'
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({ handleAddTodo }) => {
    const [{description}, handleInputChange, reset] = useForm({ //custom hook para manipular campos de formulario
        description:''
    });

    const hanleSubmit = (e) => {
        e.preventDefault();

        if(description.trim().length <= 1){ //si el campo descripcion es menor o igual a 1, se limpia espacios con trim
            return;
        }
        const newTodo = { 
            id : new Date().getTime(),
            description,
            done: false
        }

        handleAddTodo(newTodo); //llama al dispatch
        reset();//reseteo los campos del formulario
    }
    
    return (
        <>
            <h4>Agregar tarea</h4>
            <hr/>
            <form
                onSubmit={hanleSubmit}
                >
                <input
                    className="form-control"
                    type="text"
                    name="description"
                    placeholder="Indique la tarea a realizar"
                    autoComplete="off"
                    value={description}
                    onChange={handleInputChange}
                />
                <button
                    type="submit"
                    className="btn btn-primary mt-1 btn-block"
                >Agregar</button>
            </form>
        </>
    )
}
