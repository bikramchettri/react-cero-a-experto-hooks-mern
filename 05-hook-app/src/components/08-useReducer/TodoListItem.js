import React from 'react'

export const TodoListItem = ({ todo, indice, handleToggle, handleDelete}) => {
    return (
        <li
            className='list-group-item'
        >
            <p
                className={`${todo.done && 'complete'}`} /* si la tarea estÃ¡ hecha aplico la clase css complete, condicion logica and */
                onClick={() => handleToggle(todo.id)} /* para marcar como realizado */
            >
            {indice +1} {todo.description}</p>
            <button
                className="btn btn-danger"
                onClick={() => handleDelete(todo.id)} /* para borrar */
            >Borrar</button>
        </li>
    )
}
