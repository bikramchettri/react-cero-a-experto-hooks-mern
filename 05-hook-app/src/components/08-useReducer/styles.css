body{
    padding: 70px;
}
p {
    cursor: pointer;
    margin: 0px;
}
.list-group-item {
    align-items: center;
    justify-content: space-between;
    display: flex;
}

.complete {
    text-decoration: line-through;
}