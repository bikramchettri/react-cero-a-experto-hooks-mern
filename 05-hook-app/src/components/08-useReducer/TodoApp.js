import React, { useReducer, useEffect } from 'react'
import { todoReducer } from './todoReducer';
import { useForm } from '../../hooks/useForm';

import './styles.css';


const init = ( ) => {
    return JSON.parse(localStorage.getItem('todos')) || []; // regreso la lista parseada desde localstorage o un arreglo vacío
   
   /* return [{
        id : new Date().getTime(),
        description:'Aprender React',
        done: false
    }];*/
}
export const TodoApp = () => {

    const [todos, dispatch] = useReducer(todoReducer, [], init);
    
    const [{description}, handleInputChange, reset] = useForm({ //custom hook para manipular campos de formulario
        description:''
    });
    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos))
    }, [todos]);

    const handleDelete = (todoId) => {
        const action = {
            type: 'delete',
            payload: todoId
        }
        dispatch(action);//ejecuto la accion
        //reset();//reseteo los campos del formulario

    }
    const handleToggle = (todoId) => {
        
        dispatch({
            type: 'toggle',
            payload: todoId
        });
    }
    const hanleSubmit = (e) => {
        e.preventDefault();

        if(description.trim().length <= 1){ //si el campo descripcion es menor o igual a 1, se limpia espacios con trim
            return;
        }
        const newTodo = {
            id : new Date().getTime(),
            description,
            done: false
        }

        const action = {
            type: 'add',
            payload : newTodo
        }
        dispatch(action);
        reset();//reseteo los campos del formulario

    }
    return (
        <div>
            <h1>todoApp {todos.length}</h1>
            <hr/>
            <div className="row">
                <div className="col-md-7">
                    <ul className="list-group list-group-flush">
                        {
                            todos.map( (todo,indice) => (
                                <li
                                    key={todo.id}
                                    className='list-group-item'
                                >
                                <p
                                    className={`${todo.done && 'complete'}`} /* si la tarea está hecha aplico la clase css complete, condicion logica and */
                                    onClick={() => handleToggle(todo.id)} /* para marcar como realizado */
                                >
                                {indice +1} {todo.description}</p>
                                <button
                                    className="btn btn-danger"
                                    onClick={() => handleDelete(todo.id)} /* para borrar */
                                >Borrar</button>
                                </li>
                            ))
                        }

                    </ul>
                </div>
                <div className="col-md-5">
                    <h4>Agregar tarea</h4>
                    <hr/>
                    <form
                        onSubmit={hanleSubmit}
                        >
                        <input
                            className="form-control"
                            type="text"
                            name="description"
                            placeholder="Indique la tarea a realizar"
                            autoComplete="off"
                            value={description}
                            onChange={handleInputChange}
                        />
                        <button
                            type="submit"
                            className="btn btn-primary mt-1 btn-block"
                        >Agregar</button>
                    </form>
                </div>
            </div>
        </div>
    )
}
