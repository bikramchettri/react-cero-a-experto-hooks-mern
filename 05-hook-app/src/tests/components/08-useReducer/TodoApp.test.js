import React from 'react';
import {shallow, mount} from 'enzyme';
import { demoTodos } from '../../fixtures/demoTodos';
import { TodoApp } from '../../../components/08-useReducer/TodoApp';
import { act } from 'react-dom/test-utils';


describe('Pruebas en <TodoApp/>', () => {

    const wrapper = shallow(<TodoApp/>);

    Storage.prototype.setItem = jest.fn( ()=> {})

    test('Debe de mostrarse correctamente ', () => {
        expect(wrapper).toMatchSnapshot();                
    });

    test('Debe de agregar un TODO', () => {
        const wrapper =  mount(<TodoApp/>);
        act(()=> {
            wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[0]);
            wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[1]);
        });

        expect(wrapper.find('h1').text().trim()).toBe('todoApp 2')
        expect(localStorage.setItem).toHaveBeenCalledTimes(2);
        //expect(localStorage.setItem).toHaveBeenCalledWith({});
    });
    
    test('Debe de eliminar un TODO', () => {
       wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[0]); 
       wrapper.find('TodoList').prop('handleDelete')(demoTodos[0].id); 

       expect(wrapper.find('h1').text().trim()).toBe('todoApp 0')

    });
    
    
})
