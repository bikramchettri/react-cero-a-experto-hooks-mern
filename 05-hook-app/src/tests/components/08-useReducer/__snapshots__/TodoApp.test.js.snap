// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoApp/> Debe de mostrarse correctamente  1`] = `
<div>
  <h1>
    todoApp 
    0
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-md-7"
    >
      <TodoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-md-5"
    >
      <TodoAdd
        handleAddTodo={[Function]}
      />
    </div>
  </div>
</div>
`;
